cedar-lean/Cedar/Thm/Validation/Validator.lean

```diff
@@ -0,0 +1,373 @@
+/-
+ Copyright 2022-2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      https://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+-/
+
+import Cedar.Spec
+import Cedar.Data
+import Cedar.Validation
+import Cedar.Thm.Validation.Typechecker
+import Cedar.Thm.Data
+
+namespace Cedar.Thm
+
+open Cedar.Data
+open Cedar.Spec
+open Cedar.Validation
+
+/-- For a single expression, evaluates to a boolean value (or appropriate error) -/
+def OneEvaluatesCorrectly (expr : Expr) (request : Request) (entities : Entities) : Prop :=
+  ∃ (b : Bool), EvaluatesTo expr request entities b
+
+/-- Every policy as an expression evaluates to a boolean value or appropriate error -/
+def AllEvaluateCorrectly (policies : Policies) (request : Request) (entities : Entities) : Prop :=
+  ∀ policy ∈ policies, OneEvaluatesCorrectly policy.toExpr request entities
+
+def RequestAndEntitiesMatchSchema (schema : Schema) (request : Request) (entities : Entities) :Prop :=
+  ∀ env ∈ schema.toEnvironments,
+  RequestAndEntitiesMatchEnvironment env request entities
+
+def EvaluatesSubst (expr : Expr) (request : Request) (entities : Entities) : Prop :=
+  evaluate (substituteAction request.action expr) request entities = evaluate expr request entities
+
+
+theorem evaluates_subst_ite {i t e : Expr} {request : Request} {entities : Entities}
+  (ih₁ : EvaluatesSubst i request entities)
+  (ih₂ : EvaluatesSubst t request entities)
+  (ih₃ : EvaluatesSubst e request entities) :
+  evaluate (substituteAction request.action (Expr.ite i t e)) request entities =
+  evaluate (Expr.ite i t e) request entities
```

Comment:
I think it would be a little cleaner to move these proofs about `substituteAction` to their own file (e.g., `Thm/Validation/SubstituteAction.lean`)

Reply:
@emina said keeping things in one file < 1000 lines was ideal for like easy access -- any thoughts here? I would probably opt for a different file too.

Reply:
Ah, I should have clarify: we normally keep the main theorem in a separate (small) file, and all the auxiliary ones in separate (usually) big files. An example of the main theorem: https://github.com/cedar-policy/cedar-spec/blob/main/cedar-lean/Cedar/Thm/Typechecking.lean

Reply:
I don't feel strongly, so fine either way

Reply:
I think I'd personally prefer if files were capped at something more like 500 lines, not 1000 lines -- for larger files, there's probably a sensible way to split it up.  But our current codebase definitely has a bunch of large files.  Not merge blocking for sure

Reply:
I separated it out